File: client-py/src/iotdb/utils/IoTDBRpcDataSet.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

# for package
from .IoTDBConstants import *

# for debug
# from IoTDBConstants import *

import sys
from os.path import dirname, abspath
path = dirname(dirname(abspath(__file__)))
sys.path.append(path)

from thrift.transport import TTransport
from iotdb.thrift.rpc.TSIService import TSFetchResultsReq, TSCloseOperationReq


class IoTDBRpcDataSet(object):
    TIMESTAMP_STR = "Time"
    # VALUE_IS_NULL = "The value got by %s (column name) is NULL."
    START_INDEX = 2
    FLAG = 0x80

    def __init__(self, sql, column_name_list, column_type_list, column_name_index, ignore_timestamp, query_id,
        client, session_id, query_data_set, fetch_size):
        self.__session_id = session_id
        self.__ignore_timestamp = ignore_timestamp
        self.__sql = sql
        self.__query_id = query_id
        self.__client = client
        self.__fetch_size = fetch_size
        self.__column_size = len(column_name_list)
        self.__default_time_out = 1000

        self.__column_name_list = []
        self.__column_type_list = []
        self.__column_ordinal_dict = {}
        if not ignore_timestamp:
            self.__column_name_list.append(IoTDBRpcDataSet.TIMESTAMP_STR)
            self.__column_type_list.append(TSDataType.INT64)
            self.__column_ordinal_dict[IoTDBRpcDataSet.TIMESTAMP_STR] = 1

        if column_name_index is not None:
            self.__column_type_deduplicated_list = [None for _ in range(len(column_name_index))]
            for i in range(len(column_name_list)):
                name = column_name_list[i]
                self.__column_name_list.append(name)
                self.__column_type_list.append(TSDataType[column_type_list[i]])
                if name not in self.__column_ordinal_dict:
                    index = column_name_index[name]
                    self.__column_ordinal_dict[name] = index + IoTDBRpcDataSet.START_INDEX
                    self.__column_type_deduplicated_list[index] = TSDataType[column_type_list[i]]
        else:
            index = IoTDBRpcDataSet.START_INDEX
            self.__column_type_deduplicated_list = []
            for i in range(len(column_name_list)):
                name = column_name_list[i]
                self.__column_name_list.append(name)
                self.__column_type_list.append(TSDataType[column_type_list[i]])
                if name not in self.__column_ordinal_dict:
                    self.__column_ordinal_dict[name] = index
                    index += 1
                    self.__column_type_deduplicated_list.append(TSDataType[column_type_list[i]])

        self.__time_bytes = bytes(0)
        self.__current_bitmap = [bytes(0) for _ in range(len(self.__column_type_deduplicated_list))]
        self.__value = [None for _ in range(len(self.__column_type_deduplicated_list))]
        self.__query_data_set = query_data_set
        self.__is_closed = False
        self.__empty_resultSet = False
        self.__has_cached_record = False
        self.__rows_index = 0

    def close(self):
        if self.__is_closed:
            return
        if self.__client is not None:
            try:
                status = self.__client.closeOperation(TSCloseOperationReq(self.__session_id, self.__query_id))
                print("close session {}, message: {}".format(self.__session_id, status.message))
            except TTransport.TException as e:
                print("close session {} failed because: ".format(self.__session_id), e)
                raise Exception

            self.__is_closed = True
            self.__client = None

    def next(self):
        if self.has_cached_result():
            self.construct_one_row()
            return True
        if self.__empty_resultSet:
            return False
        if self.fetch_results():
            self.construct_one_row()
            return True
        return False

    def has_cached_result(self):
        return (self.__query_data_set is not None) and (len(self.__query_data_set.time) != 0)

    def construct_one_row(self):
        # simulating buffer, read 8 bytes from data set and discard first 8 bytes which have been read.
        self.__time_bytes = self.__query_data_set.time[:8]
        self.__query_data_set.time = self.__query_data_set.time[8:]
        for i in range(len(self.__query_data_set.bitmapList)):
            bitmap_buffer = self.__query_data_set.bitmapList[i]

            # another 8 new rows, should move the bitmap buffer position to next byte
            if self.__rows_index % 8 == 0:
                self.__current_bitmap[i] = bitmap_buffer[0]
                self.__query_data_set.bitmapList[i] = bitmap_buffer[1:]
            if not self.is_null(i, self.__rows_index):
                value_buffer = self.__query_data_set.valueList[i]
                data_type = self.__column_type_deduplicated_list[i]

                # simulating buffer
                if data_type == TSDataType.BOOLEAN:
                    self.__value[i] = value_buffer[:1]
                    self.__query_data_set.valueList[i] = value_buffer[1:]
                elif data_type == TSDataType.INT32:
                    self.__value[i] = value_buffer[:4]
                    self.__query_data_set.valueList[i] = value_buffer[4:]
                elif data_type == TSDataType.INT64:
                    self.__value[i] = value_buffer[:8]
                    self.__query_data_set.valueList[i] = value_buffer[8:]
                elif data_type == TSDataType.FLOAT:
                    self.__value[i] = value_buffer[:4]
                    self.__query_data_set.valueList[i] = value_buffer[4:]
                elif data_type == TSDataType.DOUBLE:
                    self.__value[i] = value_buffer[:8]
                    self.__query_data_set.valueList[i] = value_buffer[8:]
                elif data_type == TSDataType.TEXT:
                    length = int.from_bytes(value_buffer[:4], byteorder="big", signed=False)
                    self.__value[i] = value_buffer[4: 4 + length]
                    self.__query_data_set.valueList[i] = value_buffer[4 + length:]
                else:
                    print("unsupported data type {}.".format(data_type))
                    # could raise exception here
        self.__rows_index += 1
        self.__has_cached_record = True

    def fetch_results(self):
        self.__rows_index = 0
        request = TSFetchResultsReq(self.__session_id, self.__sql, self.__fetch_size, self.__query_id, True, self.__default_time_out)
        try:
            resp = self.__client.fetchResults(request)
            if not resp.hasResultSet:
                self.__empty_resultSet = True
            else:
                self.__query_data_set = resp.queryDataSet
            return resp.hasResultSet
        except TTransport.TException as e:
            print("Cannot fetch result from server, because of network connection: ", e)

    def is_null(self, index, row_num):
        bitmap = self.__current_bitmap[index]
        shift = row_num % 8
        return ((IoTDBRpcDataSet.FLAG >> shift) & (bitmap & 0xff)) == 0

    def is_null_by_index(self, column_index):
        index = self.__column_ordinal_dict[self.find_column_name_by_index(column_index)] - IoTDBRpcDataSet.START_INDEX
        # time column will never be None
        if index < 0:
            return True
        return self.is_null(index, self.__rows_index - 1)

    def is_null_by_name(self, column_name):
        index = self.__column_ordinal_dict[column_name] - IoTDBRpcDataSet.START_INDEX
        # time column will never be None
        if index < 0:
            return True
        return self.is_null(index, self.__rows_index - 1)

    def find_column_name_by_index(self, column_index):
        if column_index <= 0:
            raise Exception("Column index should start from 1")
        if column_index > len(self.__column_name_list):
            raise Exception("column index {} out of range {}".format(column_index, self.__column_size))
        return self.__column_name_list[column_index - 1]

    def get_fetch_size(self):
        return self.__fetch_size

    def set_fetch_size(self, fetch_size):
        self.__fetch_size = fetch_size

    def get_column_names(self):
        return self.__column_name_list

    def get_column_types(self):
        return self.__column_type_list

    def get_column_size(self):
        return self.__column_size

    def get_ignore_timestamp(self):
        return self.__ignore_timestamp

    def get_column_ordinal_dict(self):
        return self.__column_ordinal_dict

    def get_column_type_deduplicated_list(self):
        return self.__column_type_deduplicated_list

    def get_values(self):
        return self.__value

    def get_time_bytes(self):
        return self.__time_bytes

    def get_has_cached_record(self):
        return self.__has_cached_record

File: client-nodejs/iotdb/utils/IoTDBRpcDataSet.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {TSFetchResultsReq, TSCloseOperationReq} from "../thrift/rpc_types";
import {TSDataType} from "./IoTDBConstants";

export class IoTDBRpcDataSet {
    static TIMESTAMP_STR: string = "Time";
    static START_INDEX = 2;
    static FLAG = 0x80;

    private session_id;
    private ignore_timestamp;
    private sql;
    private query_id;
    private client;
    private fetch_size;
    private column_size;
    private default_time_out;
    private column_name_list;
    private column_type_list;
    private column_ordinal_dict;
    private column_type_deduplicated_list;
    private time_bytes;
    private current_bitmap;
    private value;
    private query_data_set;
    private is_closed;
    private empty_resultSet;
    private has_cached_record;
    private rows_index;
    private statement_id;

    constructor(
        sql,
        column_name_list,
        column_type_list,
        column_name_index,
        ignore_timestamp,
        query_id,
        client,
        statement_id,
        session_id,
        query_data_set,
        fetch_size
    ) {
        this.session_id = session_id;
        this.statement_id = statement_id;
        this.ignore_timestamp = ignore_timestamp;
        this.sql = sql;
        this.query_id = query_id;
        this.client = client;
        this.fetch_size = fetch_size;
        this.column_size = column_name_list.length;
        this.default_time_out = 1000;

        this.column_name_list = [];
        this.column_type_list = [];
        // @ts-ignore
        this.column_ordinal_dict = new Map();
        if (!ignore_timestamp) {
            this.column_name_list.push(IoTDBRpcDataSet.TIMESTAMP_STR);
            this.column_type_list.push(TSDataType.INT64);
            this.column_ordinal_dict.set(IoTDBRpcDataSet.TIMESTAMP_STR, 1);
        }

        if (column_name_index !== null) {
            this.column_type_deduplicated_list = [];
            for (let j = 0; j < column_name_index.length; j++) {
                this.column_type_deduplicated_list.push(null);
            }

            for (let i = 0; i < column_name_list.length; i++) {
                let name = column_name_list[i];
                this.column_name_list.push(name);
                this.column_type_list.push(TSDataType[column_type_list[i]]);
                if (!this.column_ordinal_dict.has(name)) {
                    let index = column_name_index[name];
                    this.column_ordinal_dict.set(name, index + IoTDBRpcDataSet.START_INDEX);
                    this.column_type_deduplicated_list[index] = TSDataType[column_type_list[i]];
                }
            }
        } else {
            let index = IoTDBRpcDataSet.START_INDEX;
            this.column_type_deduplicated_list = [];
            for (let i = 0; i < column_name_list.length; i++) {
                let name = column_name_list[i];
                this.column_name_list.push(name);
                this.column_type_list.push(TSDataType[column_type_list[i]]);
                if (!this.column_ordinal_dict.has(name)) {
                    this.column_ordinal_dict.set(name, index);
                    index += 1;
                    this.column_type_deduplicated_list.push(TSDataType[column_type_list[i]]);
                }
            }
        }

        this.time_bytes = Buffer.alloc(0);
        this.current_bitmap = [];
        for (let i = 0; i < this.column_type_deduplicated_list.length; i++) {
            this.current_bitmap.push(Buffer.alloc(0));
        }
        this.value = [];
        for (let i = 0; i < this.column_type_deduplicated_list.length; i++) {
            this.value.push(null);
        }
        this.query_data_set = query_data_set;
        this.is_closed = false;
        this.empty_resultSet = false;
        this.has_cached_record = false;
        this.rows_index = 0;
    }

    public close() {
        if (this.is_closed) {
            return;
        }
        if (this.client !== null) {
            try {
                let status = this.client.closeOperation(
                    new TSCloseOperationReq({
                        sessionId: this.session_id, 
                        queryId: this.query_id,
                        statementId: this.statement_id
                    })
                );
                console.log("close session " + this.session_id + ", message: " + status.message);
            } catch (err) {
                throw new Error("close session failed");
            }

            this.is_closed = true;
            this.client = null;
        }
    }

    public next() {
        if (this.has_cached_result()) {
            this.construct_one_row();
            return true;
        }
        if (this.empty_resultSet) {
            return false;
        }
        if (this.fetch_results()) {
            this.construct_one_row();
            return true;
        }
        return false;
    }

    public has_cached_result() {
        return (this.query_data_set !== null) && (this.query_data_set.time.length != 0);
    }

    public construct_one_row() {
        // simulating buffer, read 8 bytes from data set and discard first 8 bytes which have been read.
        this.time_bytes = this.query_data_set.time.slice(0, 8);
        let len = this.query_data_set.time.length;
        this.query_data_set.time = this.query_data_set.time.slice(8, len);
        for (let i = 0; i < this.query_data_set.bitmapList.length; i++) {
            let bitmap_buffer = this.query_data_set.bitmapList[i];

            // another 8 new rows, should move the bitmap buffer position to next byte
            if (this.rows_index % 8 == 0) {
                this.current_bitmap[i] = bitmap_buffer[0];
                let len = bitmap_buffer.length;
                this.query_data_set.bitmapList[i] = bitmap_buffer.slice(1, len);
            }
            if (!this.is_null(i, this.rows_index)) {
                let value_buffer = this.query_data_set.valueList[i];
                let data_type = this.column_type_deduplicated_list[i];

                // simulating buffer
                let len = value_buffer.length;
                if (data_type == TSDataType.BOOLEAN) {
                    this.value[i] = value_buffer.slice(0, 1);
                    this.query_data_set.valueList[i] = value_buffer.slice(1, len);
                } else if (data_type == TSDataType.INT32) {
                    this.value[i] = value_buffer.slice(0, 4);
                    this.query_data_set.valueList[i] = value_buffer.slice(4, len);
                } else if (data_type == TSDataType.INT64) {
                    this.value[i] = value_buffer.slice(0, 8);
                    this.query_data_set.valueList[i] = value_buffer.slice(8, len);
                } else if (data_type == TSDataType.FLOAT) {
                    this.value[i] = value_buffer.slice(0, 4);
                    this.query_data_set.valueList[i] = value_buffer.slice(4, len);
                } else if (data_type == TSDataType.DOUBLE) {
                    this.value[i] = value_buffer.slice(0, 8);
                    this.query_data_set.valueList[i] = value_buffer.slice(8, len);
                } else if (data_type == TSDataType.TEXT) {
                    let length = value_buffer.readInt32BE(0);
                    this.value[i] = value_buffer.slice(4, 4 + length);
                    this.query_data_set.valueList[i] = value_buffer.slice(4 + length, len);
                } else {
                    throw new Error("unsupported data type.");
                }
            }
        }
        this.rows_index += 1;
        this.has_cached_record = true;
    }

    public fetch_results() {
        this.rows_index = 0;
        let request = new TSFetchResultsReq({
            sessionId: this.session_id,
            statement: this.sql,
            fetchSize: this.fetch_size,
            queryId: this.query_id,
            isAlign: true,
            timeout: this.default_time_out
        });
        try {
            let resp = this.client.fetchResults(request);
            if (!resp.hasResultSet) {
                this.empty_resultSet = true;
            } else {
                this.query_data_set = resp.queryDataSet;
            }
            return resp.hasResultSet;
        } catch (err) {
            throw new Error("Cannot fetch result from server");
        }
    }

    public is_null(index, row_num) {
        let bitmap = this.current_bitmap[index];
        let shift = row_num % 8;
        return ((IoTDBRpcDataSet.FLAG >> shift) & (bitmap & 0xFF)) == 0;
    }

    public is_null_by_index(column_index) {
        let index = (
            this.column_ordinal_dict.get(this.find_column_name_by_index(column_index))
            - IoTDBRpcDataSet.START_INDEX
        );
        // time column will never be null
        if (index < 0) {
            return true;
        }
        return this.is_null(index, this.rows_index - 1);
    }

    public is_null_by_name(column_name) {
        let index = this.column_ordinal_dict.get(column_name) - IoTDBRpcDataSet.START_INDEX;
        // time column will never be null
        if (index < 0) {
            return true;
        }
        return this.is_null(index, this.rows_index - 1);
    }

    public find_column_name_by_index(column_index) {
        if (column_index <= 0) {
            throw new Error("Column index should start from 1");
        }
        if (column_index > this.column_name_list.length) {
            throw new Error("column index out of range");
        }
        return this.column_name_list[column_index - 1];
    }

    public get_fetch_size() {
        return this.fetch_size;
    }

    public set_fetch_size(fetch_size) {
        this.fetch_size = fetch_size;
    }

    public get_column_names() {
        return this.column_name_list;
    }

    public get_column_types() {
        return this.column_type_list;
    }

    public get_column_size() {
        return this.column_size;
    }

    public get_ignore_timestamp() {
        return this.ignore_timestamp;
    }

    public get_column_ordinal_dict() {
        return this.column_ordinal_dict;
    }

    public get_column_type_deduplicated_list() {
        return this.column_type_deduplicated_list;
    }

    public get_values() {
        return this.value;
    }

    public get_time_bytes() {
        return this.time_bytes;
    }

    public get_has_cached_record() {
        return this.has_cached_record;
    }

}
